linalg/amgxsolver.cpp

```diff
@@ -0,0 +1,880 @@
+   if (myid == 0)
+   {
+      Adisp[0] = 0;
+      for (int i=1; i<mpiTeamSz; ++i)
+      {
+         Adisp[i] = Adisp[i-1] + Apart[i-1];
+      }
+   }
+
+   MPI_Gatherv(inArr.HostRead(), inArr.Size(), MPI_INT64_T,
+               outArr.HostWrite(), Apart.HostRead(), Adisp.HostRead(),
+               MPI_INT64_T, 0, mpiTeamComm);
+
+   MPI_Barrier(mpiTeamComm);
+}
+
+void AmgXSolver::GatherArray(const Vector &inArr, Vector &outArr,
+                             const int mpiTeamSz, const MPI_Comm &mpiTeamComm,
+                             Array<int> &Apart, Array<int> &Adisp) const
+{
+   //Calculate number of elements to be collected from each process
+   int locAsz = inArr.Size();
+   MPI_Allgather(&locAsz, 1, MPI_INT,
+                 Apart.HostWrite(),1, MPI_INT, mpiTeamComm);
+
+   MPI_Barrier(mpiTeamComm);
+
+   //Determine stride for process
+   Adisp[0] = 0;
+   for (int i=1; i<mpiTeamSz; ++i)
+   {
+      Adisp[i] = Adisp[i-1] + Apart[i-1];
+   }
+
+   MPI_Gatherv(inArr.HostRead(), inArr.Size(), MPI_DOUBLE,
+               outArr.HostWrite(), Apart.HostRead(), Adisp.HostRead(),
+               MPI_DOUBLE, 0, mpiTeamComm);
+}
+
+void AmgXSolver::ScatterArray(const Vector &inArr, Vector &outArr,
+                              const int mpiTeamSz, const MPI_Comm &mpiTeamComm,
+                              Array<int> &Apart, Array<int> &Adisp) const
+{
+   MPI_Scatterv(inArr.HostRead(),Apart.HostRead(),Adisp.HostRead(),
+                MPI_DOUBLE,outArr.HostWrite(),outArr.Size(),
+                MPI_DOUBLE, 0, mpiTeamComm);
+}
+#endif
+
+void AmgXSolver::SetMatrix(const SparseMatrix &in_A, const bool update_mat)
+{
+
+   if (update_mat == false)
+   {
+
+      AMGX_matrix_upload_all(AmgXA, in_A.Height(),
+                             in_A.NumNonZeroElems(),
+                             1, 1,
+                             in_A.ReadI(),
+                             in_A.ReadJ(),
+                             in_A.ReadData(), NULL);
+
+      AMGX_solver_setup(solver, AmgXA);
+      AMGX_vector_bind(AmgXP, AmgXA);
+      AMGX_vector_bind(AmgXRHS, AmgXA);
+
+   }
+   else
+   {
+
+      AMGX_matrix_replace_coefficients(AmgXA,
+                                       in_A.Height(),
+                                       in_A.NumNonZeroElems(),
+                                       in_A.ReadData(), NULL);
+   }
+
+}
+
+#ifdef MFEM_USE_MPI
+void AmgXSolver::SetMatrix(const HypreParMatrix &A, const bool update_mat)
+{
+   hypre_ParCSRMatrix * A_ptr =
+      (hypre_ParCSRMatrix *)const_cast<HypreParMatrix&>(A);
+
+   hypre_CSRMatrix *A_csr = hypre_MergeDiagAndOffd(A_ptr);
+
+   Array<double> loc_A(A_csr->data, (int)A_csr->num_nonzeros);
+   const Array<int> loc_I(A_csr->i, (int)A_csr->num_rows+1);
+
+   //Column index must be int64_t so we must promote here
+   Array<int64_t> loc_J((int)A_csr->num_nonzeros);
+   for (int i=0; i<A_csr->num_nonzeros; ++i)
+   {
+      loc_J[i] = A_csr->big_j[i];
```

Comment:
I had to replace `big_j` with `j` to get this to compile. Does this depend on the hypre version you are using?

Reply:
I've been grabbing the version from master: `` https://github.com/hypre-space/hypre.git``. If it doesn't work with older versions I can make a note. But the column index is expected to be of int64_t type.

Reply:
`big_j` was introduced between hypre 2.15.1 and 2.16 (https://github.com/hypre-space/hypre/blob/fe8a0e0e0b1c54299f804726990d2942ae505e34/src/seq_mv/csr_matrix.h#L31). We can probably require the newer hypre for AmgX but it should be noted somewhere.

Unrelatedly, does it make sense to use `HYPRE_BigInt` instead of `int64_t`?

Reply:
I think ``HYPRE_BigInt`` has to be enabled to be int64 right? If it always is, then yes we can do that.

Reply:
We should either do some hypre version include guard here, or document carefully that our AMGx support also requires hypre >= 2.16.